### ISSUE TITLE
Windows 10 und BT Headset -> No Sounds outside of Mumble

### ISSUE THREAD

**Issue by NoobyNiceDev** [mumble-voip/mumble#3661]
Dear guys,

I'm Using Mumble just for a short time. I'm (trying) using it with a Sony WH-1000XM2 Bluetooth Headset.

My Headset ist standard Mic and Output in Windows 10 Sounds.
If I start mumble, I only can hear the mumble sounds. No other sounds are there any more. No Spotify and no system sounds.

Skype works well with my headset, also Teamspeak. So I think, Mumble is doing something different then others...

Ich think #1875 is related to this, but the workaroung mentioned there doesen't worked for me.

Also if I select my Headset in VLC as Audio Device, there is still no sound. In that moment I close mumble, Sound is back. The hole time mumble is open, windows shows this:
![grafik](https://user-images.githubusercontent.com/33202028/55577868-d7067d80-5714-11e9-9630-b3f2f40df901.png)
So Windows think, it's Playing sounds.

VLC:
![grafik](https://user-images.githubusercontent.com/33202028/55577281-83476480-5713-11e9-91d6-ce642e61aa6d.png)
![grafik](https://user-images.githubusercontent.com/33202028/55577301-95290780-5713-11e9-8bcf-4763c26f1670.png)
![grafik](https://user-images.githubusercontent.com/33202028/55577324-a6721400-5713-11e9-8518-2b28f5dbe83d.png)

**Comment by kroppt** [comment]
Try disabling 'attenuation' in the Advanced settings. It sounds as if this could be the problem. Basically, what attenuation does is lower the volume of non-Mumble applications when people are speaking in Mumble.
![image](https://user-images.githubusercontent.com/14015672/58017083-584f7f00-7acd-11e9-96c2-571086a54e74.png)
![image](https://user-images.githubusercontent.com/14015672/58017110-630a1400-7acd-11e9-9c7c-31579f5101b3.png)
![image](https://user-images.githubusercontent.com/14015672/58017130-6a312200-7acd-11e9-8d95-1b2cf1a7a57e.png)

You could also try setting both the Default Playback Device and Default Communications Device to the same output. Disabling unused devices could help keep unwanted devices inactive.
![image](https://user-images.githubusercontent.com/14015672/58017280-cc8a2280-7acd-11e9-878c-c6851b42ed5f.png)
For example, your 'Hands-Free' headphones is set to Default Communication Device, but the other device is not. Perhaps you mistook one device for the other. In my experience, if there are multiple virtual devices (Hands-Free/Stereo) installed on Windows for the same physical device (headphones), then only one of them works.

**Comment by trudnorx** [comment]
Make sure you're not using exclusive mode in Mumble. Could also try to disable exclusive mode for your sound devices in Windows.

**Comment by Krzmbrzl** [comment]
Has this issue been resolved by the suggested fixes?

**Comment by velp** [comment]
This problem still alive. I've just tried to use BT headphones Sony WH-1000XM3 but when I run Mumble all other sounds are muted.

Exclusive and Echo cancelation settings are disabled:
![image](https://user-images.githubusercontent.com/8326634/103487491-eb5e3980-4e16-11eb-9e46-221f897cc415.png)
![image](https://user-images.githubusercontent.com/8326634/103487496-f44f0b00-4e16-11eb-8626-f0f98a1715fd.png)

All other input/output devices are disabled:
![image](https://user-images.githubusercontent.com/8326634/103487508-18125100-4e17-11eb-93fa-dea73e5ebf87.png)
![image](https://user-images.githubusercontent.com/8326634/103487514-219bb900-4e17-11eb-8d88-17f1aefb8da4.png)

Mumble version:
![image](https://user-images.githubusercontent.com/8326634/103487523-311b0200-4e17-11eb-997f-8e4052321947.png)

This problem can be reproduced on two different computers with different BT drivers and as I see this problem really old: https://www.realitymod.com/forum/showthread.php?t=100715

Please take a look into it.

**Comment by timkohlman** [comment]
I have the same issue -  using Jabra Talk 45 bluetooth headset

![image](https://user-images.githubusercontent.com/22802913/103951422-c1987000-5192-11eb-8bde-b6605426a957.png)
![image](https://user-images.githubusercontent.com/22802913/103951515-edb3f100-5192-11eb-973d-3cd3583d6e1e.png)

**Comment by Ellesedil** [comment]
I'm also using a pair of Sony WH-1000XM3 headphones and also experiencing this. However, I think I have some information @velp and @timkohlman. TLDR: It's not Mumble but a quirk in how some bluetooth headsets connect to devices for high-quality sound output _or_ to support voice input/output.

Here are my sound settings:

![image](https://user-images.githubusercontent.com/9885465/104674521-a351d900-5698-11eb-9247-fd960c32d0c2.png)
![image](https://user-images.githubusercontent.com/9885465/104674754-080d3380-5699-11eb-833d-8665b604e0cd.png)
![image](https://user-images.githubusercontent.com/9885465/104674609-caa8a600-5698-11eb-8feb-65b6cfd37c01.png)

There are two sets of "default devices" here. There's the "Default Device" and the "Default Communications Device". What Mumble is doing is using the "Default Communications Device" when the Mumble settings are set to use the default Input or Output device.

For bluetooth headsets like the Sony WH-1000XM3, when you pair the headsets, you're actually pairing two bluetooth "devices" because the headsets have two different codecs it uses. There's a high-quality codec which comes across as the "WH-1000XM3 Stereo" device for output. This is used most of the time when an application doesn't connect to a microphone. And then there's a second codec, which is has lower sound quality but allows for input as well as output.

The headset isn't smart enough to use both codecs at once, so when it is connected to a program that takes voice input, the codec (and the "device") switches to the Hands-Free AG Audio device/codec. But since your programs are still trying to pipe sound to the "Stereo" device, the sound effectively dies. The sound is still being outputted to the device as far as Windows is concerned, but the headset isn't paying attention to that channel anymore, so to speak.

You thus have two options. What I did was set my default communications device to my WH-1000XM3 Stereo "device" for output, and my laptop microphone (my primary gaming rig is a laptop) for input. I had already done this for my "default device" settings, but as you can see in my screenshots, there is apparently a distinction and Mumble cares about it where Discord and other applications do not.

Or, you can force your other applications to output to the "Hands-Free AG Audio" device. I assume Jabra has a similar set-up for their bluetooth connections. This will then get them to use the same sound codec as Mumble if you have need to use the built-in microphone in the headset.

I hope this helps.

**Comment by Krzmbrzl** [comment]
@velp @timkohlman please try out the suggested workaround and report your findings. That would be very valuable information.

@Ellesedil thank you very much for sharing your insights here :+1:

**Comment by velp** [comment]
@Ellesedil thank you for explanation.

Indeed, everything works well if I use an additional microphone. For example, I have a second (wired) headphone and I can use their microphone:
![image](https://user-images.githubusercontent.com/8326634/104711671-b44f2880-5732-11eb-9ee1-fc397047b2a5.png)
![image](https://user-images.githubusercontent.com/8326634/104711693-bb763680-5732-11eb-9bc0-eb258b4f32b4.png)

But I'm trying to use exactly the same headphones for both tasks, without additional devices. So this option is not a solution to the problem.

I've tried the second way (force other applications to use Hands-Free AG Audio device) but the main problem is that I cannot choose this device as device for output in the system settings. I see it in the list of devices:
![image](https://user-images.githubusercontent.com/8326634/104712315-769ecf80-5733-11eb-8e5d-862bb34933f8.png)
but I cannot choose it:
![image](https://user-images.githubusercontent.com/8326634/104712394-8c13f980-5733-11eb-8d53-beb37043755c.png)

I found the way how to fix it: need to disable Stereo device in Control Panel
![image](https://user-images.githubusercontent.com/8326634/104725748-8f16e600-5743-11eb-99e4-e2c12511dfc9.png)

Only after that you will be able to choose "Hands-Free AG Audio" device as default output device:
![image](https://user-images.githubusercontent.com/8326634/104725839-b4a3ef80-5743-11eb-8b1e-2027591311f1.png)

So the second way fixes this problem. Also want to test it with other wireless headphones (will do it today but later) and will let you know.

**Comment by velp** [comment]
Tested with other headphones (Beats Flex) - the same solution works. Just need to disable stereo output device and choose "Hand-Free ..." device as default.
![image](https://user-images.githubusercontent.com/8326634/104726831-60017400-5745-11eb-89d3-6d26aa005682.png)

Of course, the sound quality is very bad when the stereo device is turned off, but at least there is sound.

There is only one unanswered question: why do these same headphones work correctly in other programs (like Discord, Google meet, Telegram, Slack) without disabling drivers and such settings?

**Comment by Krzmbrzl** [comment]
At least we now know how to make it work :+1: 

> There is only one unanswered question: why do these same headphones work correctly in other programs (like Discord, Google meet, Telegram, Slack) without disabling drivers and such settings?

This is an interesting question indeed.

In the screenshots it is seen that the headset registers as two different audio devices and Ellesedil mentioned that Mumble is using the default communication device by default. Is it perhaps possible to select the stereo device from within Mumble directly? :thinking:

**Comment by velp** [comment]
Yes, I can enable both devices:
![image](https://user-images.githubusercontent.com/8326634/104769178-f488c800-577f-11eb-8950-f139a2f0224e.png)

and choose the stereo device in mumble as output device and as default device in the system:
![image](https://user-images.githubusercontent.com/8326634/104769232-0b2f1f00-5780-11eb-928c-7e41c967860d.png)

but in this case sounds from other application do not work.

Only if I disable stereo device and choose "Hand-Free ..." device in the system and in the Mumble I can hear the sounds from other applications. 

BTW in Discord when I go to the page with sound setting and testing:
![image](https://user-images.githubusercontent.com/8326634/104769699-c35cc780-5780-11eb-88ac-107cfcf4b28b.png)

I see the same issue. But only on this page, because Discord in this page takes exclusive mode for sound devices. I know this because volume level contol blocked in the system:
![image](https://user-images.githubusercontent.com/8326634/104770616-e3d95180-5781-11eb-9e1e-6b84a05319b1.png)

The same situation I see when I run Mumble:
![image](https://user-images.githubusercontent.com/8326634/104770820-3adf2680-5782-11eb-9a44-25dff2f34105.png)

but also there are two Mumbles.

So it looks like Mumble just ignores disabled "exclusive" mode.

**Comment by Krzmbrzl** [comment]
@davidebeatrici do you know how the exclusive mode stuff is handled on Windows?

**Comment by davidebeatrici** [comment]
We explicitly ask Windows to open the device for us in exclusive mode:
https://github.com/mumble-voip/mumble/blob/2fdd96b20088e62dc4a565ba16197f64216da7e4/src/mumble/WASAPI.cpp#L444-L491

**Comment by Krzmbrzl** [comment]
But this seems like we are respecting the `exclusive` setting. Therefore it should only be opened in this way, if the user chooses to do so, right? :thinking:

**Comment by davidebeatrici** [comment]
Indeed.

**Comment by timkohlman** [comment]
Thanks for the info @Ellesedil      yeah I set the Headset to "Default Device" and it worked ok. 

before: 
![image](https://user-images.githubusercontent.com/22802913/105017754-425f4200-5a98-11eb-99dc-4e787eb792a5.png)

after:
![image](https://user-images.githubusercontent.com/22802913/105017884-6de22c80-5a98-11eb-9774-278cd6f8c079.png)

**Comment by mirh** [comment]
Forget the exclusive setting, as already said this happens because the activation of the microphone triggers the switch to the HFP profile, while "normal audio" is still sent over A2DP. 
You can trigger the same just by opening the "Recording" tab of the OS audio proprieties window. 

> There is only one unanswered question: why do these same headphones work correctly in other programs (like Discord, Google meet, Telegram, Slack) without disabling drivers and such settings?

Idk about all the electron shitte, but I just tested telegram and it's the same. You are probably hitting https://github.com/telegramdesktop/tdesktop/issues/9814

**Comment by Krzmbrzl** [comment]
In that case I don't think we can do something against that. Thus I am closing this issue for now.
